File: test/performance.jl
using CSV
using DataFrames
using BenchmarkTools

import RecurrenceAnalysis as RA
import RecurrenceMicrostatesAnalysis as RMA

function rma_test(data)
    det = RMA.determinism(data, 0.27)[1]
    lam = RMA.laminarity(data, 0.27)[1]

    return det, lam
end

function ra_test(data)
    rp = RA.RecurrenceMatrix(data, 0.27)
    det = RA.determinism(rp)
    lam = RA.laminarity(rp)

    return det, lam
end

rng_df = CSV.read("range.csv", DataFrame)
rng = rng_df[!, 1]

results_ra = zeros(Float64, length(rng))
results_rma = zeros(Float64, length(rng))

data_gl = (CSV.read("perf_data/db_1_500.csv", DataFrame))[!, :]

for i in eachindex(rng)
    data_df = CSV.read("perf_data/db_$(i)_$(rng[i]).csv", DataFrame)
    global data_gl = data_df[!, 1]

    bm_ra = @benchmark ra_test(data_gl)
    bm_rma = @benchmark rma_test(data_gl)
    

    results_ra[i] = mean(bm_ra.times) / 1e9
    results_rma[i] = mean(bm_rma.times) / 1e9
end

df_results = DataFrame(rng = rng, rma_res = results_rma, ra_res = results_ra)
CSV.write("benchmark_jl_results.csv", df_results)

File: test/performance.py
import timeit
import numpy as np
import pandas as pd
import recurrence_microstates as rma
from pyunicorn.timeseries import RecurrencePlot

##
##      Benchmark test using RMA
def rma_test(data):
    det = rma.determinism(data, 0.27)
    lam = rma.laminarity(data, 0.27)

    return det, lam

##
##      Benchmark test using PyUnicorn
def unicorn_test(data):
    rp = RecurrencePlot(data, threshold = 0.27, normalize = False, v_min = 2, l_min = 2)
    det = rp.determinism()
    lam = rp.laminarity()

    return det, lam

##
##      Load range
df = pd.read_csv("range.csv")
rng = df["rng"].values

##
##      Alocate memory for our results.
results_pyunicorn = np.zeros(len(rng))
results_rma = np.zeros(len(rng))

for i in range(len(rng)):
    df_data = pd.read_csv(f"perf_data/db_{i + 1}_{rng[i]}.csv")
    data = df_data["data"].values
    
    def wrapper_rma(data = data):
        return rma_test(data)
    
    def wrapper_pyunicorn(data = data):
        return unicorn_test(data)
    
    tests_unicorn = np.zeros(40)
    tests_rma = np.zeros(40)

    for j in range(40):
        tests_unicorn[j] = timeit.timeit(wrapper_pyunicorn, number = 1)
        tests_rma[j] = timeit.timeit(wrapper_rma, number = 1)
    
    results_pyunicorn[i] = np.mean(tests_unicorn)
    results_rma[i] = np.mean(tests_rma)

df_results = pd.DataFrame({
    "rng": rng,
    "rma_time": results_rma,
    "pyunicorn_time": results_pyunicorn
})
df_results.to_csv("benchmark_py_results.csv", index = False)